### PR TITLE
Added in code to process Wikidata, generate gold and negative links

### PR DESCRIPTION
**Pull request by anthonywchen**
Added code into `retrieval_augmented_generation` to process Wikidata, generate gold links, and generate negative links

### CODE REVIEW

**Comment by anthonywchen** [comment]
Resolves issues #1 and #2

**Comment by anthonywchen** [comment]
@samanz-verneek

**Review by samanz-verneek — COMMENTED** [review]
Code looks pretty good. Not urgent, but these are the kind of methods that would benefit from testing.

File: retrieval_augmented_linking/generate_data/extract_entity_relations.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import tqdm
+import ujson as json
+
+# pylint: disable=c-extension-no-member
+
+
+def dumb_filter(line):
+    """Does a simple check on a Wikidata line that is a dictionary in string format."""
+    # Filter entities that don't have an English Wikipedia page
+    if "\"enwiki\"" not in line and "\"type\":\"item\"" in line:
```
[the remark this answers is not in the document]
Comment (anthonywchen) [reply retrieval_augmented_linking/generate_data/extract_entity_relations.py:25]: good point.

File: retrieval_augmented_linking/generate_data/extract_entity_relations.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            if entry['mainsnak']['datatype'] == 'wikibase-item' and entry['mainsnak']['snaktype'] == 'value':
+                # If the tail entity is in the list of filtered entities, keep it
+                tail_entity = entry['mainsnak']['datavalue']['value']['id']
+                if tail_entity in filtered_entities:
+                    relations[relation_id].append(tail_entity)
+
+    return relations
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (anthonywchen) [reply retrieval_augmented_linking/generate_data/extract_entity_relations.py:47]: yeah that's a good idea. I'll leave this to until the end though. If you have a repo where you have an example of this that'd be good.

File: retrieval_augmented_linking/generate_data/extract_entity_relations.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    then write this information into a JSON file. We write each dictionary of entity
+    relations in it's own line for easy readability.
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--wikidata_dump", help=".json Wikidata dump for information extraction")
+    parser.add_argument("--filtered_entities", help="JSON file for containing filtered Wikidata entity")
+    parser.add_argument("--output_file", help="Output JSON file for writing Wikidata entity relations")
+    args = parser.parse_args()
+
+    with open(args.filtered_entities) as reader:
```
[the remark this answers is not in the document]
Comment (anthonywchen) [reply retrieval_augmented_linking/generate_data/extract_entity_relations.py:59]: I don't quite follow. Does adding an argument count as business logic? This argument seems pretty business agnostic

File: retrieval_augmented_linking/generate_data/extract_entity_relations.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        writer.write("{\n")
+
+        with open(args.wikidata_dump) as reader:
+            # Each line corresponds to a dictionary about a Wikidata entity
+            for line_num, line in enumerate(tqdm.tqdm(reader, desc="Extracting relations")):
+                if dumb_filter(line):
+                    continue
+
+                try:
+                    line = json.loads(line.strip()[:-1])
```
[the remark this answers is not in the document]
Comment (anthonywchen) [reply retrieval_augmented_linking/generate_data/extract_entity_relations.py:72]: Ah no. This is a bit of a hack I should fix. The first and last lines of the file are "{" and "}" respectively so I should just use an if/else block here.

File: retrieval_augmented_linking/generate_data/extract_entity_relations.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (samanz-verneek) [reply retrieval_augmented_linking/generate_data/extract_entity_relations.py:59]: Sorry, I think this wasn't clear. This reading code fit's better in a method that main calls. For instance, if later we want to be able to call that from a library we can easily do that. With the current setup, if we want to read as a library, we'd have to call `main`.

File: retrieval_augmented_linking/generate_data/extract_entity_relations.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (samanz-verneek) [reply retrieval_augmented_linking/generate_data/extract_entity_relations.py:72]: That's fine. Python style seems to like these try, except blocks over if/else in ways that I've never been too comfortable with coming from other langs. I think a comment instead about why you use it is fine.

File: retrieval_augmented_linking/generate_data/generate_gold_links.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+import tqdm
+import ujson as json
+
+random.seed(0)
+
+# pylint: disable=c-extension-no-member, duplicate-code
+
+
+def generate_link_dict(entity_information, head_entity, relation, tail_entity):
```
[the remark this answers is not in the document]
Comment (anthonywchen) [reply retrieval_augmented_linking/generate_data/generate_gold_links.py:49]: Yup good point.

File: retrieval_augmented_linking/generate_data/filter_entities.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+import argparse
+import collections
+
+import tqdm
+import ujson as json
+
+# pylint: disable=c-extension-no-member
+
+
+def dumb_filter(line):
```
[the remark this answers is not in the document]
Comment (anthonywchen) [reply retrieval_augmented_linking/generate_data/filter_entities.py:23]: it is, I should abstract and make it's own utils.py file

File: retrieval_augmented_linking/real/dataset.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        num_docs,
+        num_subvectors=128,
+        num_probes=5,
+        hnsw_m=32,
+        nbits=8,
+        metric=faiss.METRIC_INNER_PRODUCT
+    ):
+        super().__init__()
+        nlist = math.floor(math.sqrt(num_docs))
+        quantizer = faiss.IndexHNSWFlat(embed_dim, hnsw_m)
```
Comment (samanz-verneek) [inline retrieval_augmented_linking/real/dataset.py:125]: Can you comment on what these are and why used?

File: train_real.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+from pytorch_lightning.callbacks import ModelCheckpoint
+from pytorch_lightning.loggers import TensorBoardLogger
+
+from retrieval_augmented_linking.real.dataset import REALDataset
+from retrieval_augmented_linking.real.model import REAL
+
+
+class REALCallback(pl.Callback):
+    """
+    Save a checkpoint every N steps, instead of Lightning's default that checkpoints
```
Comment (samanz-verneek) [inline train_real.py:19]: Seems like this callback does more than this? Also, wondering why all this needs to be a callback? can't some be in the LightningModule?

File: train_real.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+                        input_ids=torch.stack(input_ids[i:i+bs]).to(model.device),
+                        attention_mask=torch.stack(attention_mask[i:i+bs]).to(model.device)
+                    ).pooler_output.detach().cpu().numpy())
+                # Get `num_knn` document IDs of KNN of links in batch
+                _, nn = dataset.faiss.search(embeds, top_k=num_knn)
+                # Update KNN of batch
+                dataset.knn[i:i+embeds.shape[0]] = nn
+            model.train()
+
+    def save_checkpoint(self, trainer):
```
Comment (samanz-verneek) [inline train_real.py:126]: Maybe the checkpoint logic should be it's own callback?

**Review by samanz-verneek — CHANGES_REQUESTED** [review]

**Review by giannisp-verneek — CHANGES_REQUESTED** [review]

File: evaluate_real.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    nearest_neighbors_list = []
+    for i in tqdm.tqdm(range(0, len(input_ids), bs), desc="Updating KNN of links"):
+        with torch.no_grad():
+            embeds = np.ascontiguousarray(model.query_encoder(
+                input_ids=torch.stack(input_ids[i:i+bs]).to(model.device),
+                attention_mask=torch.stack(attention_mask[i:i+bs]).to(model.device)
+            ).pooler_output.detach().cpu().numpy())
+        nearest_neighbors_list += index.search(embeds, k=args.num_neighbors)[1].tolist()
+    assert len(nearest_neighbors_list) == len(link_ids_list)
+    assert len(nearest_neighbors_list[0]) == args.num_neighbors
```
[the remark this answers is not in the document]
Comment (samanz-verneek) [reply evaluate_real.py]: What's wrong with assert here?

**Review by giannisp-verneek — APPROVED** [review]

**Pull request merged**